src/lasse_handler.erl

```diff
@@ -20,186 +21,190 @@
-    module() |
-    #{ module => module()
-     , init_args => any()
-     }.
-
--spec init(any(), cowboy_req:req(), lasse_handler_options()) ->
-    {loop, any(), state()}.
-init(Transport, Req, []) ->
-    init(Transport, Req, #{});
-init(Transport, Req, [Module]) when is_atom(Module) ->
-    init(Transport, Req, #{module => Module});
-init(_Transport, Req, Opts) ->
-    try
-      #{module := Module} = Opts,
-      InitArgs = maps:get(init_args, Opts, []),
-      {LastEventId, Req} = cowboy_req:header(<<"last-event-id">>, Req),
-      InitResult = Module:init(InitArgs, LastEventId, Req),
-      handle_init(InitResult, Module)
-    catch
-      _:{badmatch, #{}} ->
-        throw(module_option_missing)
-    end.
+  module() |
+  #{ module => module()
+   , init_args => any()
+   }.
+
+-spec init(cowboy_req:req(), lasse_handler_options()) ->
+  {ok | module(), any(), state()}.
+init(Req, []) ->
+    init(Req, #{});
+init(Req, [Module]) when is_atom(Module) ->
+    init(Req, #{module => Module});
+init(Req, Opts) ->
+  try
+    #{module := Module} = Opts,
+    InitArgs = maps:get(init_args, Opts, []),
+    LastEventId = cowboy_req:header(<<"last-event-id">>, Req),
+    InitResult = Module:init(InitArgs, LastEventId, Req),
+    handle_init(InitResult, Module)
+  catch
+    _:{badmatch, #{}} ->
+      throw(module_option_missing)
+  end.
 
 -spec info(term(), cowboy_req:req(), state()) ->
-    {ok|loop, cowboy_req:req(), state()}.
+    {ok|stop, cowboy_req:req(), state()}.
 info({message, Msg}, Req, State) ->
-    Module = State#state.module,
-    ModuleState = State#state.state,
-    Result = Module:handle_notify(Msg, ModuleState),
-    process_result(Result, Req, State);
+  Module = State#state.module,
+  ModuleState = State#state.state,
+  Result = Module:handle_notify(Msg, ModuleState),
+  process_result(Result, Req, State);
 info(Msg, Req, State) ->
-    Module = State#state.module,
-    ModuleState = State#state.state,
-    Result = Module:handle_info(Msg, ModuleState),
-    process_result(Result, Req, State).
+  Module = State#state.module,
+  ModuleState = State#state.state,
+  Result = Module:handle_info(Msg, ModuleState),
+  process_result(Result, Req, State).
 
 -spec terminate(term(), cowboy_req:req(), state()) -> ok.
-terminate(Reason, Req, State = #state{}) ->
-    Module = State#state.module,
-    ModuleState = State#state.state,
-    Module:terminate(Reason, Req, ModuleState),
-    ok.
+terminate(Reason, Req, State) ->
+  Module = State#state.module,
+  ModuleState = State#state.state,
+  Module:terminate(Reason, Req, ModuleState),
+  ok.
+
+-spec upgrade(Req, Env, module(), any(), timeout(), run | hibernate)
+  -> {ok, Req, Env} | {suspend, module(), atom(), [any()]}
+  when Req::cowboy_req:req(), Env::cowboy_middleware:env().
+upgrade(Req, Env, Handler, HandlerState, Timeout, run) ->
+  cowboy_loop:upgrade(Req, Env, Handler, HandlerState, Timeout, run);
+upgrade(Req, Env, Handler, HandlerState, Timeout, hibernate) ->
```

Comment:
Why 2 clauses if you're basically doing the same in both? Why not just:
```erlang
upgrade(Req, Env, Handler, HandlerState, Timeout, Action) ->
  cowboy_loop:upgrade(Req, Env, Handler, HandlerState, Timeout, Action).
```

you could've even just _mixed_ the function _in_ using mixer.